File: src/main/scala/org/joshjoyce/lumivore/io/DirectoryPathStream.scala
package org.joshjoyce.lumivore.io

import java.io.File
import java.nio.file.{AccessDeniedException, Files, Path}
import scala.collection.JavaConversions

object DirectoryPathStream {
  def recurse(root: File)(f: Path => Any) = {
    val d = new DirectoryPathStream(root, f)
    d.start()
  }
}

class DirectoryPathStream(val dir: File, visitor: Path => Any) {
  import JavaConversions._

  def start() {
    if (dir.isFile) {
      visitor(dir.toPath)
    } else {
      try {
        val children = Files.newDirectoryStream(dir.toPath)

        children.foreach {
          c => {
            val s = new DirectoryPathStream(c.toFile, visitor)
            s.start()
          }
        }
      } catch {
        case (e: AccessDeniedException) => {}// skip it
        case (e: Exception) => throw e
      }
    }
  }
}

File: src/main/java/org/joshjoyce/lumivore/io/DirectoryPathStream.java
package org.joshjoyce.lumivore.io;

import java.io.File;
import java.nio.file.*;
import java.util.function.Consumer;

public class DirectoryPathStream {
    private final File dir;
    private final Consumer<Path> visitor;

    public DirectoryPathStream(File dir, Consumer<Path> visitor) {
        this.dir = dir;
        this.visitor = visitor;
    }

    public static void recurse(File root, Consumer<Path> f) {
        var d = new DirectoryPathStream(root, f);
        d.start();
    }

    public void start() {
        if (dir.isFile()) {
            visitor.accept(dir.toPath());
        } else {
            try {
                var children = Files.newDirectoryStream(dir.toPath());
                children.forEach(
                    c -> {
                        var s = new DirectoryPathStream(c.toFile(), visitor);
                        s.start();
                    }
                );
            } catch (AccessDeniedException e) {
                // skip it
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }
    }
}
